src/circuit/gadget/sinsemilla/commit_ivk.rs

```diff
@@ -0,0 +1,802 @@
+use halo2::{
+    circuit::Layouter,
+    plonk::{Advice, Column, ConstraintSystem, Error, Expression, Selector},
+    poly::Rotation,
+};
+use pasta_curves::{arithmetic::FieldExt, pallas};
+
+use crate::{
+    circuit::gadget::{
+        ecc::{chip::EccChip, X},
+        utilities::{bitrange_subset, bool_check, copy, CellValue, Var},
+    },
+    constants::T_P,
+};
+
+use super::{
+    chip::{SinsemillaChip, SinsemillaCommitDomains, SinsemillaConfig},
+    CommitDomain, Message, MessagePiece,
+};
+
+#[derive(Clone, Debug)]
+pub struct CommitIvkConfig {
+    q_canon: Selector,
+    advices: [Column<Advice>; 10],
+    sinsemilla_config: SinsemillaConfig,
+}
+
+impl CommitIvkConfig {
+    pub(in crate::circuit) fn configure(
+        meta: &mut ConstraintSystem<pallas::Base>,
+        advices: [Column<Advice>; 10],
+        sinsemilla_config: SinsemillaConfig,
+    ) -> Self {
+        let q_canon = meta.selector();
+
+        let config = Self {
+            q_canon,
+            advices,
+            sinsemilla_config,
+        };
+
+        // <https://zips.z.cash/protocol/nu5.pdf#concretesinsemillacommit>
+        // We need to hash `ak || nk` where each of `ak`, `nk` is a field element (255 bits).
+        //
+        // a = bits 0..=249 of `ak`
+        // b = b_0||b_1||b_2`
+        //   = (bits 250..=253 of `ak`) || (bit 254 of  `ak`) || (bits 0..=4 of  `nk`)
+        // c = bits 5..=244 of `nk`
+        // d = d_0||d_1` = (bits 245..=253 of `nk`) || (bit 254 of `nk`)
+        //
+        // `a`, `b`, `c`, `d` have been constrained by the Sinsemilla hash to be:
+        //   - a: 250 bits,
+        //   - b: 10 bits,
+        //   - c: 240 bits,
+        //   - d: 10 bits
+        //
+        meta.create_gate("CommitIvk canonicity check", |meta| {
```

Comment:
Currently this gate uses the following region layout over 10 advice columns:

![image](https://user-images.githubusercontent.com/4993799/126713714-d43aa583-f6ca-4e51-9e4d-3af4a592aa0e.png)

I would prefer that we use the following layout over 9 advice columns:

![image](https://user-images.githubusercontent.com/4993799/126713926-ada3f252-7c70-429e-ab32-d17173c7f090.png)